README.rst
=================
kombinat.decogrid
=================


Introduction
============

``kombinat.decogrid`` is a css code generator for layouts with `Deco Grid System`_ based.
It generates css code similar to "`columns.css`_" in `plonetheme.sunburst`_.
Copy the generated css into ``your.product/skins/prod_styles/columns.css`` to
override any other definitions.


Features
========

- Provides the generates css code similar to `plonetheme.sunburst`_ ``columns.css`` file.


Installation
============

If you are a developer, you might enjoy installing it via buildout.

For install ``kombinat.decogrid`` package add it to your ``buildout`` section's 
*eggs* parameter e.g.: ::

    [buildout]
    parts =
        ...
        decogrid

    [decogrid]
    recipe = zc.recipe.egg
    eggs = kombinat.decogrid


and then running ``bin/buildout``.

Or, you can add it as a dependency on your own product ``setup.py`` file: ::

    install_requires=[
        ...
        'kombinat.decogrid',
    ],


Guide
=====

The autogenerated executable is located in your buildout ``bin/`` directory.
It allows you several options for generating decogrid css code, which can be
directly pasted into your ``columns.css`` file::

    ~/buildout $ bin/decogrid_generatecss -h
    usage: decogrid_generatecss [-h] [-c COLUMNS] [-m MARGIN] [-w WIDTH]
                                [-l LEVEL] [-t TAG] [-o]

    decogrid_generatecss [options] decogrid_generatecss generate decogrid
    columns.css for plone (http://deco.gs) Default values are written in brackets.

    optional arguments:
      -h, --help            show this help message and exit
      -c COLUMNS, --columns COLUMNS
                            how many columns (16)
      -m MARGIN, --margin MARGIN
                            margin in pixels between each cell (10)
      -w WIDTH, --width WIDTH
                            portal width in pixels including potentially left
                            and/or right margins (960)
      -l LEVEL, --convenience-level LEVEL
                            max level of convenience classes to create (7)
      -t TAG, --tag TAG     html tag to be used for 'row', 'cell' and
                            'convenience' css classes (div)
      -o, --omit-margin     omit left and right margin around the portal
                            (recommended for nesting grids inside grids)


Example
-------

This utility is very useful if you want to use nested grid layout because you
can omit the margin which is given by default in plone's ``columns.css``.

Generate a new ``columns.css`` file by using:

::

    ./bin/decogrid_generatecss -o -w 950 -c 16 > columns.css


Contribute
==========

- Issue Tracker: https://github.com/collective/kombinat.decogrid/issues
- Source Code: https://github.com/collective/kombinat.decogrid


License
=======

The project is licensed under the GPLv2.


Credits
-------

- Daniel Widerin (daniel at widerin dot net).


Amazing contributions
---------------------

- Leonardo J. Caballero G. aka macagua (leonardocaballero at gmail dot com).

You can find an updated list of package contributors on https://github.com/collective/kombinat.decogrid/contributors

.. _`Deco Grid System`: http://deco.gs
.. _`columns.css`: https://github.com/plone/plonetheme.sunburst/blob/master/plonetheme/sunburst/skins/sunburst_styles/columns.css
.. _`plonetheme.sunburst`: https://github.com/plone/plonetheme.sunburst


CHANGES.rst
Changelog
=========

0.4 (unreleased)
----------------

- Changed the Deco.gs from the script command description
  [macagua]

- Changed the i18n domain for grid-test page template
  [macagua]

- Added improvements about README file
  [macagua]

- Added more strings classifiers items for the metadata package
  [macagua]

- Fixed warning: no files found matching '*.txt' from MANIFEST.in file
  [macagua]

- Fix packaging error, ``CHANGES.rst`` was missing in eggs.
  [saily]


0.3 (2013-10-09)
----------------

- Switch from ``optparse`` to ``argparse``, so we need >= Python 2.7 now.
  [saily]

- Add more configurable parser, allow non-div tags, allow to strip tags by
  passing ``""`` as argument for ``-t``.
  [saily]

- Allow users to limit convenience classes by using ``--level`` parameter.
  default is still ``min(columns, 7)``.
  [saily]


0.2 - 2012-07-09
----------------

- fixed css lineendings
- removed double "position-0" declaration
  [petschki]

0.1.1 - 2011-05-26
------------------

- updated install documentation [petschki]

0.1 - 2011-04-07
----------------

- Idea and initial release
  [saily]


kombinat/decogrid/grid.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import argparse
import os.path
import string
import sys


class Deco(object):

    css_row = "row"
    css_cell = "cell"
    css_width = "width"
    css_position = "position"

    template = string.Template(open(
        os.path.join(os.path.dirname(__file__), 'deco.css'), 'r').read())

    def __init__(self, width, margin, columns, omit, maxlevel=7, tag='div'):
        self.width = width
        self.margin = margin
        self.columns = columns
        self.omit = omit
        self.maxlevel = maxlevel
        self.tag = tag

    @property
    def p_cell(self):
        return 100.0 / self.width * self.cell

    @property
    def p_margin(self):
        return 100.0 / self.width * self.margin

    @property
    def portal_margin(self):
        if self.omit:
            return 0
        return self.margin / 2.0

    @property
    def p_portal_margin(self):
        return 100.0 / self.width * self.portal_margin

    @property
    def cell(self):
        if self.omit:
            margin = (self.columns - 1) * self.margin
        else:
            margin = self.columns * self.margin

        return float(self.width - margin) / self.columns

    @property
    def width_full(self):
        if self.omit:
            return 100

        return 100.0 / self.width * (self.width - self.margin)

    @property
    def position_0(self):
        if self.omit:
            return - 100

        return 100.0 / self.width * (self.width - (self.margin * 0.5))

    def widths(self):
        css = ""
        for i in range(1, self.columns + 1):
            w = 100.0 / self.width * (i * self.cell + (i - 1) * self.margin)
            css += "\n%s.%s-%d { width: %.4f%%; }" % (
                self.tag, self.css_width, i, w)

        return css

    def positions(self):
        css = ""
        for i in range(0, self.columns):
            if self.omit:
                p = -100 + 100.0 / self.width * \
                    i * (self.cell + self.margin)
            else:
                p = -100 + 100.0 / self.width * \
                    (i * self.cell + (i + 0.5) * self.margin)

            css += "\n%s.%s-%d { margin-left: %.4f%%; }" % (
                self.tag, self.css_position, i, p)

        return css

    def conveniences(self):
        css = """\
%s.%s-full { width: %.4f%%; }
""" % (self.tag, self.css_width, self.width_full)

        for i in range(2, min(self.maxlevel, self.columns)):
            pos = ""
            for j in range(1, i):

                if self.omit:
                    cell = float(self.width - (i - 1) * self.margin) / i
                else:
                    cell = float(self.width - self.margin * i) / i

                css += "\n%s.%s-%d\\3a %d { width: %.4f%%; }" % (
                    self.tag, self.css_width, j, i,
                    100.0 / self.width * (j * cell + (j - 1) * self.margin))

                pos += "\n%s.%s-%d\\3a %d { margin-left: %.4f%%; }" % (
                    self.tag, self.css_position, j, i, -100 + (
                        100.0 / self.width * (j * cell + (j + (
                            not self.omit and 0.5 or 0)) * self.margin)))

            css += pos + "\n"

        return css

    def __getitem__(self, key):
        value = getattr(self, key)
        if callable(value):
            return value()

        return value

    def __call__(self):
        if int(self.cell) != self.cell:
            sys.stdout.write("/* WARN: You don't have int pixel values.*/\n\n")

        return self.template.substitute(self)


def generator():
    parser = argparse.ArgumentParser(description="""%(prog)s [options]

%(prog)s generate decogrid columns.css for Plone (https://github.com/limi/deco.gs) Default \
values are written in brackets.""")

    parser.add_argument('-c', '--columns', dest='columns', default=16,
                        type=int, metavar="COLUMNS",
                        help="how many columns (16)")

    parser.add_argument('-m', '--margin', dest='margin', default=10,
                        type=int, metavar="MARGIN",
                        help="margin in pixels between each cell (10)")

    parser.add_argument('-w', '--width', dest='width', default=960,
                        type=int, metavar="WIDTH",
                        help="portal width in pixels including potentially " +
                             "left and/or right margins (960)")

    parser.add_argument('-l', '--convenience-level', type=int, dest='level',
                        default=7,
                        help="max level of convenience classes to create (7)")

    parser.add_argument('-t', '--tag', type=str, dest='tag', default='div',
                        help="html tag to be used for 'row', 'cell' and "
                             "'convenience' css classes (div)")

    # boolean parameters
    parser.add_argument('-o', '--omit-margin', action='store_true',
                        dest='omit', default=False,
                        help="omit left and right margin around the portal " +
                             "(recommended for nesting grids inside grids)")

    options = parser.parse_args()

    deco = Deco(options.width,
                options.margin,
                options.columns,
                options.omit,
                options.level,
                options.tag)

    sys.stdout.write(deco())


if __name__ == '__main__':
    generator()
